### ISSUE TITLE
Delivery Report Callback Queue

### ISSUE THREAD

**Issue by jarreds** [edenhill/librdkafka#242]
I noticed that delivery reports are called by popping messages off of a queue [here](https://github.com/edenhill/librdkafka/blob/master/src/rdkafka.c#L1664-L1702).

The application I'm building needs to insert checkpoint entries with the producer interspersed among the actual messages. I do not want to checkpoint after every message successfully produced. I can implement this in a number of ways, but...

It would be ideal if I could obtain info related to this dr callback queue, or at least know that another callback is going to be invoked without a call to ```rd_kafka_poll```. In that case, I would hold off on checkpointing until I knew that the delivery report callback queue is drained. Perhaps there is another mechanism to do this that I haven't discovered. Your feedback would be great.

*An even crazier request that should be another issue... it would be amazing if there were a file descriptor on the delivery callbacks so I could do io multiplexing via ```select``` with some other network sockets the app waits on.*

Thank you

**Comment by jarreds** [comment]
Also, fwiw, my use case would be significantly simplified if https://github.com/edenhill/librdkafka/issues/204 were implemented. Especially if batch ordering was guaranteed.

**Comment by edenhill** [comment]
**Checkpointing**
`rd_kafka_poll()` will serve the entire queue in one call, and it returns the number of events processed (`dr_cb, error_cb, stats_cb`).
So if your use case allows it you could do the checkpointing after `rd_kafka_poll()` returns with a positive value (better yet, have a counter in your dr_cb to rule out any error or stats callbacks).

**Regarding fd based polling**
Yeah, this is the common problem with a lack of unified polling system, there are sockets and there are thread-queues, and they are not easily polled together.
What you could do is create a separate thread that feeds `rd_kafka_poll()` events into a file descriptor and then poll that fd from your main loop.


**#204**
By batch ordering, do you mean guaranteed ordering within a batch (intra), or guaranteed ordering between batches (inter)? (B1 is delivered before B2).
The latter can only be supported if the number of in-flight simultaneous Produce requests is limited to 1, and that will effect performance (thruput will be batchsize / broker_roundtrip).

**Comment by jarreds** [comment]
I feel silly that I didn't think of checkpointing after polling along with a combination of collecting ```dr_cb``` info. That definitely makes the most sense. Thank you.

Re: polling. It was a long-shot that you had something like this already implemented in librdkafka. I'll take a stab at the thread based signaling. Thank you for the suggestion.

Re: 204. I mean intra-batch ordering. My use case is an application that reads a socket of another systems write-ahead log data. The application then replicates the log information to Kafka.  I was interested in the batch produce because it would be fairly simple to collect data on the socket up to a batch size, or until no more data exists, and then issue a produce request. The application would then wait for the produce to succeed, and then continue again, doing the same thing. In this case, batch size is a function of how much data is on the socket. Which is exactly what I want. The atomicity guarantees are very important since this is essentially a replication mechanism. Currently, if delivering a message fails for any reason, the application starts over at the last known *sequentially* delivered message, and re-produces messages again from that point on.

**Comment by edenhill** [comment]
Great!

**204**
Okay, intra-batch ordering and atomicity will be guaranteed in #204.

**Comment by jarreds** [comment]
Wonderful. Has work started on the implementation of the batch API yet?

Also, I'm 100% willing and able to help out if you have any tasks I can help tackle... coding, testing, whatever.

**Comment by edenhill** [comment]
No, work has not started yet unfortunately.

Great, more hands and eyes are always welcome!  Are you heavier on the C or C++ side?

**Comment by edenhill** [comment]
See #204
